Custom matrix lite support for split keyboards

jaygreco opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description
<!--- Describe your changes in detail here. -->
Until recently, split boards with a custom matrix needed to use `CUSTOM_MATRIX = yes` rather than the simpler `lite` option.
The new unified split common + regular matrix introduced in https://github.com/qmk/qmk_firmware/pull/13330 is really close to supporting `CUSTOM_MATRIX = lite`. If the debounce code is omitted from `quantum/matrix_common/matrix_scan()` when the `SPLIT_KEYBOARD` and `CUSTOM_MATRIX` options are set, the lite option can be used. 

Tested with split board that will be added in a follow-on PR and also on a NIBBLE, which uses `CUSTOM_MATRIX = lite` but is not a split board in order to make sure the changes didn't break existing non-split boards.

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

drashna commented on quantum/matrix_common.c:
```diff
@@ -104,16 +104,19 @@ __attribute__((weak)) void matrix_init(void) {
         matrix[i]     = 0;
     }
 
+#ifndef SPLIT_KEYBOARD
     debounce_init(MATRIX_ROWS);
+#endif
```
```suggestion
    debounce_init(ROWS_PER_HAND);
```

And add this at the top:
```c
#ifdef SPLIT_KEYBOARD
#    define ROWS_PER_HAND (MATRIX_ROWS / 2)
#else
#    define ROWS_PER_HAND (MATRIX_ROWS)
#endif
```

drashna commented on quantum/matrix_common.c:
```diff
@@ -104,16 +104,19 @@ __attribute__((weak)) void matrix_init(void) {
     bool changed = matrix_scan_custom(raw_matrix);
 
+#ifndef SPLIT_KEYBOARD
     debounce(raw_matrix, matrix, MATRIX_ROWS, changed);
-
+#endif
```
same here. 
```suggestion
    debounce(raw_matrix, matrix, ROWS_PER_HAND, changed);
```

drashna reviewed (commented):
seems to compile as is. 

Also, it may make more sense to move even more of the split code into matrix_common, so that matrix.c is mostly *just* handling the matrix read, and matrix_common is doing all the heavy lifting?

jaygreco commented:
Great suggestion. I agree. I was considering that but was apprehensive about moving too much functionality across without discussing it. I'll make and test that change & update the PR.

jaygreco commented:
Alright, I tested the changes as suggested. The order of the debounce matters for split keyboards. The debounce routine needs to be run before the slave matrix is read in, or it will double-debounce and toggle held keys each read, which is what I'm seeing. 

quantum/matrix.c:362
```c
#ifdef SPLIT_KEYBOARD
    debounce(raw_matrix, matrix + thisHand, ROWS_PER_HAND, changed);
    changed = (changed || matrix_post_scan());
#else
    debounce(raw_matrix, matrix, ROWS_PER_HAND, changed);
    matrix_scan_quantum();
#endif
```

Any thoughts?

jaygreco commented:
@drashna I pushed a commit that takes a hack at this. If nothing else, it's a starting point. It adds the debouncing in the right sequence. It also adds a new weak function prototype for `matrix_post_scan_custom`. A question that's probably worth asking is if it's worth it (and/or appropriate) to move the standard `matrix_post_scan` over to `matrix_common` entirely, since right now my `matrix_post_scan_custom` is just a carbon copy of that code.

I also made a gist that has the meat of my `custom_matrix` function just to add some context: https://gist.github.com/jaygreco/56319ffc5b15a567f125fabfff796649

jaygreco commented on quantum/matrix_common.c:
```diff
@@ -104,15 +120,25 @@ __attribute__((weak)) void matrix_init(void) {
     bool changed = matrix_scan_custom(raw_matrix);
 
-    debounce(raw_matrix, matrix, MATRIX_ROWS, changed);
+#ifdef SPLIT_KEYBOARD
```
I don't love doing this with an #ifdef, but IMO doing something like setting `thisHand` to 0 and passing that for a non-split seems like it would be confusing.

drashna commented:
The post scan function should be probably moved wholesale, without any custom option, as its sole purpose is to handle the transport code, and doesn't actually touch the matrix, directly.

jaygreco commented:
Shuffled things to make matrix_post_scan live in matrix_common, which cleans things up considerably on the [keyboard side](https://gist.github.com/jaygreco/56319ffc5b15a567f125fabfff796649).

drashna commented:
Merge conflicts.

jaygreco commented:
Checking in on this (I thought it had been merged before the breaking changes but was mistaken). If I resolve the conflicts, can it be merged into the dev branch?

drashna commented:
This is cause errors during compile: 
```
build_keyboard.mk:461: *** [.build/obj_handwired_tractyl_manuform_5x6_right_f411_drashna/quantum/matrix_common.o] error 1

#0  .build/obj_handwired_tractyl_manuform_5x6_right_f411_drashna/quantum/process_keycode/process_tap_dance.o at /Users/drashna/repos/qmk_firmware/build_keyboard.mk:457
#1  .build/handwired_tractyl_manuform_5x6_right_f411_drashna.elf at /Users/drashna/repos/qmk_firmware/tmk_core/rules.mk:362
#2  elf at /Users/drashna/repos/qmk_firmware/tmk_core/rules.mk:283
#3  build at /Users/drashna/repos/qmk_firmware/build_keyboard.mk:445
#4  all at /Users/drashna/repos/qmk_firmware/build_keyboard.mk:439
Compiling: quantum/matrix_common.c                                                                 In file included from quantum/logging/debug.h:21,
                 from quantum/keymap.h:34,
                 from quantum/quantum.h:21,
                 from quantum/matrix_common.c:1:
quantum/matrix_common.c: In function 'matrix_print':
quantum/logging/print.h:85:39: error: unknown conversion type character 'b' in format [-Werror=format=]
   85 | #define print_bin_reverse8(i) xprintf("%08b", bitrev(i))
      |                                       ^~~~~~
quantum/matrix_common.c:63:35: note: in expansion of macro 'print_bin_reverse8'
   63 | #    define print_matrix_row(row) print_bin_reverse8(matrix_get_row(row))
      |                                   ^~~~~~~~~~~~~~~~~~
quantum/matrix_common.c:81:9: note: in expansion of macro 'print_matrix_row'
   81 |         print_matrix_row(row);
      |         ^~~~~~~~~~~~~~~~
quantum/logging/print.h:85:43: note: format string is defined here
   85 | #define print_bin_reverse8(i) xprintf("%08b", bitrev(i))
      |                                           ^
quantum/logging/print.h:85:39: error: too many arguments for format [-Werror=format-extra-args]
   85 | #define print_bin_reverse8(i) xprintf("%08b", bitrev(i))
      |                                       ^~~~~~
quantum/matrix_common.c:63:35: note: in expansion of macro 'print_bin_reverse8'
   63 | #    define print_matrix_row(row) print_bin_reverse8(matrix_get_row(row))
      |                                   ^~~~~~~~~~~~~~~~~~
quantum/matrix_common.c:81:9: note: in expansion of macro 'print_matrix_row'
   81 |         print_matrix_row(row);
      |         ^~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
 [ERRORS]
`
